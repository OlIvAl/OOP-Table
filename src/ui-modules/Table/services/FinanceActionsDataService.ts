import { action, observable, runInAction } from 'mobx';
import { IBalanceParams, IFinanceActionsDataService } from '../state/interfaces';
import {
  IBalanceStateData,
  IFinanceActionData,
  IFinanceActionTotal
} from '../../../DataLayer/modules/actions/interfaces';
import DataLayer from '../../../DataLayer';

export class FinanceActionsDataService implements IFinanceActionsDataService {
  @observable loading: boolean = true;
  @observable actions: IFinanceActionData[] = [];
  @observable total: IFinanceActionTotal | null = null;
  @observable totalAmount: number = 0;

  @action.bound async *init(params: IBalanceParams): AsyncGenerator<void> {
    yield this.setLoading();

    yield this.getData(params);

    yield this.unsetLoading();
  }

  /**
   * Делает запрос за данными списка букингов
   * Инициализирует хранилище списка букингов
   * Инициализирует хранилище програм для фильтрации
   * @param params Объект с данными для запроса букингов
   */
  @action.bound protected async getData(params: IBalanceParams): Promise<void> {
    try {
      const data: IBalanceStateData = await DataLayer.getBalanceStateData(params);

      runInAction(() => {
        this.actions = data.actions;
        this.total = data.total_row;
        this.totalAmount = data.total;
        this.actions = data.actions;
        this.total = data.total_row;
        this.totalAmount = data.total;
      });
    } catch (e) {
      throw e;
    }
  }

  /**
   * Ставит флаг загрузки в true
   */
  @action.bound protected setLoading(): void {
    this.loading = true;
  }
  /**
   * Ставит флаг загрузки в false
   */
  @action.bound protected unsetLoading(): void {
    this.loading = false;
  }
}
