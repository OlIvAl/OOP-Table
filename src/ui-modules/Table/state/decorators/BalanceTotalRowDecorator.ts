import { IBalanceTotalRow } from '../interfaces';
import { IFinanceActionTotal } from '../../../../DataLayer/modules/actions/interfaces';
import { FormatHelper } from '../../../../lib/FormatHelper';

export class BalanceTotalRowDecorator implements IBalanceTotalRow {
  source: '' = '';
  created_at: '' = '';
  updated_at: '' = '';
  description: '' = '';
  action_state: '' = '';

  get price(): string {
    if (!this.wrappee.price) {
      return 'â€”';
    }
    return FormatHelper.formatCellValue(this.wrappee.price, 'money', this.currency) as string;
  }

  get total_profit(): string {
    return FormatHelper.formatCellValue(this.wrappee.total_profit, 'money', this.currency) as string;
  }

  get profit_diff(): string {
    return FormatHelper.formatCellValue(this.wrappee.profit_diff, 'money', this.currency) as string;
  }

  protected wrappee: IFinanceActionTotal;
  protected currency: string;

  constructor(wrappee: IFinanceActionTotal | null, currency: string) {
    if (!wrappee) {
      throw Error('Total field must not be null');
    }

    this.wrappee = wrappee;
    this.currency = currency;
  }
}
