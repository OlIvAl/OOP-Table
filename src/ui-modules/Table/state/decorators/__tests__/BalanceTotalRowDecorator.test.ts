import { BalanceTotalRowDecorator } from '../BalanceTotalRowDecorator';
import { FinanceActionsDataService } from '../../../services/FinanceActionsDataService';
import { flow } from 'mobx';
import { IBalanceTotalRow, IFinanceActionsDataService } from '../../interfaces';

const currency = 'rub';
let financeActionsDataService: IFinanceActionsDataService;
let totalRow: IBalanceTotalRow;

beforeEach(async () => {
  financeActionsDataService = new FinanceActionsDataService();
  await flow(financeActionsDataService.init)({});

  totalRow = new BalanceTotalRowDecorator(financeActionsDataService.total, currency);
});

describe('BalanceTotalRowDecorator', () => {
  describe('constructor', () => {
    it('init valid object', () => {
      expect(totalRow).toMatchSnapshot();
    });
    it('throw error, if model is null', () => {
      expect(() => new BalanceTotalRowDecorator(null, currency)).toThrowError(/.+/);
    });
  });
  describe('totalRow', () => {
    it('get right totalRow', () => {
      const expectedBalanceTotalRow: IBalanceTotalRow = {
        total_profit: '0 ₽',
        profit_diff: '1 000 ₽',
        created_at: '',
        description: '',
        price: '100 ₽',
        source: '',
        action_state: '',
        updated_at: ''
      };
      expect(totalRow.total_profit).toEqual(expectedBalanceTotalRow.total_profit);
      expect(totalRow.profit_diff).toEqual(expectedBalanceTotalRow.profit_diff);
      expect(totalRow.created_at).toEqual(expectedBalanceTotalRow.created_at);
      expect(totalRow.description).toEqual(expectedBalanceTotalRow.description);
      expect(totalRow.price).toEqual(expectedBalanceTotalRow.price);
      expect(totalRow.source).toEqual(expectedBalanceTotalRow.source);
      expect(totalRow.action_state).toEqual(expectedBalanceTotalRow.action_state);
      expect(totalRow.updated_at).toEqual(expectedBalanceTotalRow.updated_at);
    });
  });
  describe('totalRow with null price', () => {
    it('get right totalRow with "-" on the price place', () => {
      financeActionsDataService.total.price = null;
      totalRow = new BalanceTotalRowDecorator(financeActionsDataService.total, currency);

      const expectedBalanceTotalRow: IBalanceTotalRow = {
        total_profit: '0 ₽',
        profit_diff: '1 000 ₽',
        created_at: '',
        description: '',
        price: '—',
        source: '',
        action_state: '',
        updated_at: ''
      };

      expect(totalRow.total_profit).toEqual(expectedBalanceTotalRow.total_profit);
      expect(totalRow.profit_diff).toEqual(expectedBalanceTotalRow.profit_diff);
      expect(totalRow.created_at).toEqual(expectedBalanceTotalRow.created_at);
      expect(totalRow.description).toEqual(expectedBalanceTotalRow.description);
      expect(totalRow.price).toEqual(expectedBalanceTotalRow.price);
      expect(totalRow.source).toEqual(expectedBalanceTotalRow.source);
      expect(totalRow.action_state).toEqual(expectedBalanceTotalRow.action_state);
      expect(totalRow.updated_at).toEqual(expectedBalanceTotalRow.updated_at);
    });
  });
});
