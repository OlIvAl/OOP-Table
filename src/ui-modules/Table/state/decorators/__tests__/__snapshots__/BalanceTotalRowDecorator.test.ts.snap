// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BalanceTotalRowDecorator constructor init valid object 1`] = `
BalanceTotalRowDecorator {
  "action_state": "",
  "created_at": "",
  "currency": "rub",
  "description": "",
  "source": "",
  "updated_at": "",
  "wrappee": Object {
    "price": 100,
    "profit_diff": 1000,
    "total_profit": 0,
  },
}
`;
